---
title: "2D Analysis Notebook"
output:
  html_document: 
    toc: yes
    theme: united
    fig_width: 10
    fig_height: 10
    fig_caption: yes
---

```{r}
library(tidyverse)
library(CCA)
library(readxl)
```

# Read in the Datasets
```{r}
UpGene <- read.csv("~/Documents/AlinaRnaSeq/IvC/UpGenelist_infectedvscontrol.csv")

DownGene <- read.csv("~/Documents/AlinaRnaSeq/IvC/DownGenelist_infected_infectedvscontrol.csv")
```

## Log2FC
```{r}
l2fctable_InfectedvsControl_BL6 <- read.csv("~/2D_Analysis/l2fctable_InfectedvsControl_BL6.csv")

# filter the symbol column based on only the genes that are in UpGenes so that these are both significant and differentially regulated

l2fctable_InfectedvsControl_BL6 <- filter(l2fctable_InfectedvsControl_BL6,
                                      symbol %in% c(UpGene$x, DownGene$x))
# l2fctable_InfectedvsControl_BL6 <- filter(l2fctable_InfectedvsControl_BL6, symbol %in% UpGene$x)
```

```{r}
rownames(l2fctable_InfectedvsControl_BL6) <- l2fctable_InfectedvsControl_BL6[, 2] # make the symbols as rownames
l2fctable_InfectedvsControl_BL6 <- subset(l2fctable_InfectedvsControl_BL6, select = -c(X, symbol)) # remove the X and symbol columns
net_l2fc <- rowSums(l2fctable_InfectedvsControl_BL6, na.rm = TRUE, dims = 1)
# sum up the values in each row for every gene to get net infected gene expression
l2fctable_InfectedvsControl_BL6$net_l2fc <- net_l2fc
```

```{r}
library(stringr)
colnames(l2fctable_InfectedvsControl_BL6) <- str_remove_all(colnames(l2fctable_InfectedvsControl_BL6),
  pattern = "log2FoldChange_"
)
l2fctable_InfectedvsControl_BL6 <- l2fctable_InfectedvsControl_BL6[order(l2fctable_InfectedvsControl_BL6$net_l2fc, decreasing = TRUE), ]
head(l2fctable_InfectedvsControl_BL6)
```

## Effector Table

```{r}
effector <- as.data.frame(read_excel("~/2D_Analysis/allStrains_PresentAbsent_effectorlist_forKeshav.xlsx",
  col_types = c(
    "text", "numeric", "numeric", "numeric", "numeric",
    "numeric", "numeric", "numeric", "numeric", "numeric",
    "numeric", "numeric", "numeric", "numeric", "numeric"
  )
))
rownames(effector) <- effector[, 1] # move effectors to rownames
effector <- subset(effector, select = -...1) # remove effector symbol column
net_eff <- rowSums(effector, na.rm = TRUE, dims = 1)
effector$net_effector <- net_eff
head(effector)
```





```{r}
mc <- matcor(t(l2fctable_InfectedvsControl_BL6), t(effector))
correlate <- mc$XYcor
```
```{r fig.height=15, fig.width=15}
library(corrplot)
# corrplot(correlate, method = "color", type = "full", is.corr = TRUE, diag = TRUE)
```


```{r}
library(corrr)
corr_df <- as.data.frame(correlate) %>% fashion()
```

```{r}
# library(CCA)
# cc_results <- cor(l2fc_matrix, effectot_matrix)
```

```{r}
# x <- cor(l2fc_matrix, effector_matrix,method = c("spearman"))
#
# out <- data.frame(MaxCorr = apply(x,1,max), T_ColIndex=apply(x,1,which.max),C_ColIndex=1:nrow(x))
#
# head(out)
```
```{r}
# cancor(l2fc$net_l2fc, effector$net_effector)
```
```{r}
# library(GGally)
# ggpairs(l2fc)
```


R Canonical Correlation Analysis
```{r}
# library(corrr)
# cc1 <- cc(l2fc, effector)
```
```{r}
# compute canonical loadings
# cc2 <- comput(l2fc, effector, cc1)
```


```{r}
effnew <- read.csv("~/2D_Analysis/eff.csv")
rownames(effnew) <- effnew[, 1] # move effectors to rownames
effnew <- subset(effnew, select = - X) # remove effector symbol column
head(effnew)
```


# Factor Analysis for Mixed Data - FAMD

```{r}
library("FactoMineR")
library("factoextra")
```

```{r}
#l2fc <- l2fctable_InfectedvsControl_BL6[ c(1:28, 392:419), 1:14]
l2fc <- l2fctable_InfectedvsControl_BL6[c(1:20 , 220:240), 1:14]
#eff <- effector[1:6, 1:14]
l2fc[is.na(l2fc)] <- 0.0000001
```

```{r fig.height=15, fig.width=10}
library("gplots")

# 1. convert the data as a table
l2fc_matrix <- as.table(as.matrix(l2fc))
# 2. Graph
balloonplot(t(l2fc_matrix), main = "l2fc", xlab = "", ylab = "",label.lines = T,
            label = FALSE, show.margins = F)
```
```{r}
# eff[eff == '1'] <- 'Present'
# eff[eff == '2'] <- 'Present'
# eff[eff == '0'] <- 'Absent'

l <- t(l2fc)
e <- t(effnew)

df_merge <- merge(l, e, by = 'row.names', all = TRUE)
row.names(df_merge) <- df_merge$Row.names
df_merge <- subset(df_merge, select = -Row.names)

head(df_merge)
```

```{r}
res.famd <- FAMD(df_merge, graph = F)

print(res.famd)
```
```{r}
summary(res.famd)
```

## Eigen Value

Eigenvalues correspond to the amount of information retained by each axis. Dimensions are ordered decreasingly and listed according to the amount of variance explained in the solution. Dimension 1 explains the most variance in the solution, followed by dimension 2 and so on.

```{r}
eig.val <- get_eigenvalue(res.famd)
head(eig.val)
```


### Visualize eigenvalues (scree plot). 

Show the percentage of variances explained by each principal component.
```{r}
barplot(eig.val[, 2], 
        names.arg = 1:nrow(eig.val), 
        main = "Variances Explained by Dimensions (%)",
        xlab = "Principal Dimensions",
        ylab = "Percentage of variances",
        col = "steelblue")
# Add connected line segments to the plot
lines(x = 1:nrow(eig.val), eig.val[, 2], 
      type = "b", pch = 19, col = "red")
```
## Variables
```{r}
var <- get_famd_var(res.famd)
var
```
### Contribution of Variables to dimensions

```{r}
var$contrib
```
## Graph of variables for different dimensions

```{r fig.align='default', fig.width=10}
# Plot of variables
fviz_famd_var(res.famd, 
              repel = TRUE, 
              geom = c("point", "text"),
              axes = c(1,2),
              col.ind = "green",
              col.var.sup = "green"
              )
```
```{r fig.width=10}
# Plot of variables
fviz_famd_var(res.famd, repel = TRUE, axes = c(2,3))
```


```{r fig.width=10}
# Plot of variables
fviz_famd_var(res.famd, repel = TRUE, axes = c(3,4))
```
```{r fig.width=10}
# Plot of variables
fviz_famd_var(res.famd, repel = TRUE, axes = c(4,5))
```


## Contributions Plots

This function can be used to visualize the contribution of rows/columns from the results of FAMD.

```{r}
# Contribution to the first dimension
fviz_contrib(res.famd, "var", axes = 1, sort.val = "desc", top = 40)
# Contribution to the second dimension
fviz_contrib(res.famd, "var", axes = 2, sort.val = "desc", top = 40)
```
The red dashed line on the graph above indicates the expected average value, If the contributions were uniform. 

## Quantitative variables

```{r}
quanti.var <- get_famd_var(res.famd, "quanti.var")
quanti.var 
```

```{r}
fviz_famd_var(res.famd, "quanti.var", repel = TRUE)
```
Briefly, the graph of variables (correlation circle) shows the relationship between variables, the quality of the representation of variables, as well as, the correlation between variables and the dimensions. 

The most contributing quantitative variables can be highlighted on the scatter plot using the argument col.var = "contrib". This produces a gradient colors, which can be customized using the argument gradient.cols.

```{r}
fviz_famd_var(res.famd, "quanti.var", col.var = "contrib", 
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE)
```

Similarly, you can highlight quantitative variables using their cos2 values representing the quality of representation on the factor map. If a variable is well represented by two dimensions, the sum of the cos2 is closed to one. For some of the items, more than 2 dimensions might be required to perfectly represent the data.

```{r}
# Color by cos2 values: quality on the factor map
fviz_famd_var(res.famd, "quanti.var", 
              col.var = "cos2",
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"), 
             repel = TRUE,
             
             )
```
## Graph of qualitative variables

```{r}
quali.var <- get_famd_var(res.famd, "quali.var")
quali.var 
```
### Qualitative Variables Categories

The plot below shows the relationships between row points (Qualitative Variables):

1. Rows with a similar profile are grouped together.
2. Negatively correlated rows are positioned on opposite sides of the plot origin (opposed quadrants).
3. The distance between row points and the origin measures the quality of the row points on the factor map. 

```{r fig.height=8, fig.width=12}
fviz_famd_var(res.famd, "quali.var", col.var = "contrib",
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07")
             )
```


## Graph of individuals

Individuals with similar profiles are close to each other on the factor map. Qualitative variable categories are shown in black.
```{r}
ind <- get_famd_ind(res.famd)
ind
```

## Biplot/Symmetric Plot

1. The graph below is called symmetric plot and shows a global pattern within the qualitative data. 
2. Rows are represented by blue points and columns by red triangles.
3. The distance between any row points or column points gives a measure of their similarity (or dissimilarity). 
4. Row points with similar profile are closed on the factor map. The same holds true for column points.
5. Symetric plot represents the row and column profiles simultaneously in a common space. In this case, only the            distance between row points or the distance between column points can be really interpreted.

6. The distance between any row and column items is not meaningful! One can only make a general statements about the       observed pattern.

```{r}
fviz_hmfa_quali_biplot(res.famd, 1:2, repel = TRUE, 
                       )
```

```{r fig.width=10}
fviz_famd_ind(res.famd, 
              col.ind = "contrib",
              gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
              repel = TRUE)
```

### Correlation between variables and principal dimensions (qualitative and quantitative)
```{r fig.height =10, fig.width=10}
plot(res.famd, choix = "var")
```



## Dimension Description
```{r}
res.desc <- dimdesc(res.famd, axes = 1:5, proba = 0.05)

```
```{r}
# Description of dimension 1
res.desc[[1]]
```
```{r}
# Description of dimension 2
res.desc[[2]]
```
```{r}
# Description of dimension 3
res.desc[[3]]
```
```{r}
# Description of dimension 4
res.desc[[4]]
```
```{r}
# Description of dimension 5
res.desc[[5]]
```



