---
title: "Lets Go: From Bams to Count Matrix Notebook"
output: html_notebook
---

```{r}
library(Rsubread)
```

# Feature counts
```{r}
# GTF File
gtf_file <- "C:/Users/User/Documents/bam/Mus_musculus.GRCm39.106.chr.gtf"

```

```{r}
# Bam files
bam_path <- "C:/Users/User/Documents/bam/"
bamFiles <- list.files(bam_path, pattern = ".*bam$")
bamFiles
```


```{r}
setwd(bam_path)
fc <- featureCounts(files = bamFiles,
                    annot.ext= gtf_file,
                    isGTFAnnotationFile=TRUE,
                    isPairedEnd=FALSE)
```



```{r}
fc
```

# Differential Gene Expression Analysis

```{r}
save(fc, file = "C:/Users/User/Documents/bam/featurecounts.csv")
```

```{r}
library("DESeq2")
library("PoiClaClu")
library("pheatmap")
library("RColorBrewer")
library('tidyverse')
library("PoiClaClu")
library("vsn")
library('EnhancedVolcano')
```

```{r}
fc$target
```

