---
title: ' Differential Gene Expression Analysis Notebook'
output:
  pdf_document: 
    toc: yes
    number_sections: yes
  html_document:
    toc: yes
    fig_width: 10
    fig_height: 10
    fig_caption: yes
---

```{r}
library("DESeq2")
library("PoiClaClu")
library("pheatmap")
library("RColorBrewer")
library('tidyverse')
library("PoiClaClu")
library("vsn")
library('EnhancedVolcano')
library('gplots')
library('org.Mm.eg.db')
library('stringr')
library("genefilter")
library("dplyr")
library("ggplot2")
library("glmpca")
library('org.Mm.eg.db')
library("AnnotationDbi")
library("apeglm")
library("ComplexHeatmap")
library("clusterProfiler")
library('ggrepel')
library('corrplot')
library("GO.db")
library('GOstats')
library('pathview')
library("gage")
library("gageData")
library('GOSemSim')
library('DOSE')
library('enrichplot')
library('ggnewscale')
```

# Differential Gene Expression Analysis

## Creating metadata for the DGE Analysis

DESeq2 needs sample information (metadata) for performing DGE analysis.
Let's create the sample information

```{r}
# Read the csv file and change the column name 
sample_ID <- read.csv("~/R/Alina_RNAseq/samples.csv")
head(sample_ID)
```

```{r}
condition <- c("Infected", "Infected", "Infected", "Infected", "Infected", "Infected",
               "Infected", "Infected", "Infected", "Infected", "Infected", "Infected",
               "Infected", "Infected", "control", "control")
coldata <- data.frame(sample_ID, condition)
colnames(coldata) <- c('Sample_Name','condition') # change name of one of the columns
# The metadata can be found in a df called coldata!
head(coldata)
```

### Tidying up the names for plots later!

#### First from coldata

```{r}
# tidying up the names od samples in both columns that list of samples
#coldata$Samples <- str_remove_all(coldata$Sampls, pattern = "run6_trimmed_|_.bam|_S\\d\\d|_S\\d")
coldata$Sample_Name <- str_remove_all(coldata$Sample_Name,
                                  pattern = "run6_trimmed_|_.bam|_S\\d\\d|_S\\d" )
coldata$condition <- as.factor(coldata$condition)
```

### Changing the names of samples (as per Alina)
```{r}
coldata[coldata == '476_R1'] <- 'T'
coldata[coldata == '754_R1'] <- 'S54'
coldata[coldata == '755_R1'] <- 'S55'
coldata[coldata == '757_R1'] <- 'L57'
coldata[coldata == '758_R1'] <- 'A58'
coldata[coldata == '760_R1'] <- 'L60'
coldata[coldata == '761_R1'] <- 'S61'
coldata[coldata == '762_R1'] <- 'A62'
coldata[coldata == '763_R1'] <- 'L63'
coldata[coldata == '764_R1'] <- 'A64'
coldata[coldata == '765_R1'] <- 'S65'
coldata[coldata == '766_R1'] <- 'L66'
coldata[coldata == '768_R1'] <- 'A68'
coldata[coldata == '769_R1'] <- 'L69'
coldata[coldata == 'Ctrl1_R1'] <- 'C1'
coldata[coldata == 'Ctrl2_R2'] <- 'C2'
# convert column1 with sample names to row.names of coldata
rownames(coldata) <- coldata$Sample_Name
coldata
```
## Adding the groupings by Alina for further Metadata Information

```{r}
coldata$Epithelial_response <- c ("LowInducer", "LowInducer", "HighInducer",
                                  "HighInducer", "LowInducer", "LowInducer",
                                  "HighInducer", "HighInducer", "LowInducer",
                                  "HighInducer", "LowInducer", "HighInducer",
                                  "LowInducer", "LowInducer", 'NR', 'NR')
coldata$clinical_outcome <- c('symptomatic', 'symptomatic', 'symptomatic',
                              'Lethal', 'asymptomatic', 'Lethal', 'symptomatic',
                              'asymptomatic', 'Lethal', 'symptomatic', 'symptomatic',
                              'Lethal', 'asymptomatic', 'Lethal', 'NR', 'NR')
coldata$microcolonies <- c('Low', 'Low', 'Low', 'High', 'Low', 'Low',
                           'High', 'High', 'Low', 'High', 'Low', 'High', 'Low',
                           'Low', 'NR', 'NR')
coldata$ER_microcolonies <- c("LI_LM", "LI_LM", "HI_LM", "HI_HM", "LI_LM", "LI_LM",
                              "HI_HM", "HI_HM", "LI_LM", "HI_HM", "LI_LM", "HI_HM",
                              "LI_LM", "LI_LM", 'NR', 'NR')
coldata$phylogenomic_lineage <- c("EPEC1", "EPEC10", "EPEC9", "EPEC9", "NC", "EPEC5",
                                  "EPEC8", "NC", "EPEC7", "NC", "EPEC2", "EPEC9",
                                  "EPEC2", "EPEC2", 'NR', 'NR')
coldata$phylogroup <- c("B2", "A", "B2", "B2", "B1", "A", "B2", "B2", "B1", "B2", "B1",
                        "B2", "B1", "B2", 'NR', 'NR')
coldata$Intimin_Type <- c("alpha", "ND", "lambda", "lambda", "epsilon", "epsilon",
                          "mu", "lambda", "beta", "kappa", "beta", "alpha", "beta",
                          "beta", 'NR', 'NR')
```

#### then fix Countsmatrix:

NOTE:

1.  From the manuals the countsData must be a numeric matrix
2.  It is IMPORTANT to keep the names of the genes in the rownames

```{r}
# Readin  countsmatrix
#countsmatrix <-as.matrix(read.csv("~/R/Rtuts/Data/Alina_EPEC_project/counts.csv"))
countsmatrix <- read.csv("~/R/Alina_RNAseq/newcounts.csv")
#countsmatrix <- as.data.frame(countsmatrix)
```

```{r}
## Removal of Gender Genes from ENSEMBL ID itself
countsmatrix <- countsmatrix %>% filter(countsmatrix$X != "ENSMUSG00000086503",
                                  countsmatrix$X != "ENSMUSG00000097571",
                                  countsmatrix$X != "ENSMUSG00000086370",
                                  countsmatrix$X != "ENSMUSG00000031329")
nrow(countsmatrix)
countsmatrix <- as.matrix(countsmatrix)
```
```{r}
#tidying up these names again
#colnames(countsmatrix) <- str_remove_all(colnames(countsmatrix), pattern = "run6_trimmed_|_.bam|_S\\d\\d|_S\\d")
rownames(countsmatrix) <- countsmatrix[,1] #converting first column of gene names into rownames, to be used for sanity check later
# It is IMPORTANT to keep the names of the genes in the rownames
countsmatrix <- subset(countsmatrix, select = - X)#dropping the X column
# the elements from Sample_Name from coldata must the the colnames of countsmatrix
colnames(countsmatrix) <- coldata$Sample_Name
# Display the column names
colnames(countsmatrix)
# Convert the countsmatrix elements to be of numeric type in order to be in correct format to be fed into DESeq2 functions
class(countsmatrix) <- "numeric"
```
# Differential Gene Expression analysis using DESeq2

Now, construct DESeqDataSet for DGE analysis.

But before that, a sanity check : It is essential to have the name of
the columns in the count matrix in the same order as that in name of the
samples (rownames in coldata).

```{r}
all(rownames(coldata) %in% colnames(countsmatrix))
ncol(countsmatrix) == nrow(coldata)
dim(countsmatrix)
```

## Creating the DESeq Data set Object

```{r}
dds_infected <- DESeqDataSetFromMatrix(countData = countsmatrix,
                              colData = coldata, 
                              design = ~ condition)
nrow(dds_infected)
```

## Exploratory Data Analysis and Visualization

### Pre-filtering the dataset

Our count matrix with our DESeqDataSet contains many rows with only zeros, and additionally many rows with only a few
fragments total. In order to reduce the size of the object, and to increase the speed of our functions, we can remove
the rows that have no or nearly no information about the amount of gene expression.

Applying the most minimal filtering rule: removing rows of the DESeqDataSet that have no counts, or only a single count
across all samples. Additional weighting/filtering to improve power is applied at a later step in the workflow.

```{r}
keep <- rowSums(counts(dds_infected)) > 1
dds_infected <- dds_infected[keep,]
nrow(dds_infected)
```
### The variance stabilizing transformation

## Applying VST transformation

```{r}
vsd <- vst(dds_infected, blind = FALSE)
head(assay(vsd), 3)
colData(vsd)
vsd_coldata <- colData(vsd)
```

```{r}
dds_infected <- estimateSizeFactors(dds_infected)
print(dds_infected)
```
## Sample Distances

useful first step in an RNA-seq analysis is often to assess overall
similarity between samples:

1.  Which samples are similar to each other, which are different?
2.  Does this fit to the expectation from the experiment's design?

### Euclidean Distance between samples

dist to calculate the Euclidean distance between samples - useful for
ONLY normalized data. To ensure we have a roughly equal contribution
from all genes, we use it on the VST data.

```{r}
sampleDists <- dist(t(assay(vsd)))
head(sampleDists)
```

visualize the distances in a heatmap

```{r}
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- vsd$Samples 
colnames(sampleDistMatrix) <- vsd$Samples
```
```{r}
colors <- colorRampPalette( rev(brewer.pal(9, "RdYlBu")) )(255)
distance_plot <- pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         main = "Sample-to-Sample Euclidean Distance of BL6 - Infected Vs Control",
         col = colors)
distance_plot
```

### Poisson Distance between Samples
```{r}
poisd <- PoissonDistance(t(counts(dds_infected))) # raw counts or unnormalised data
samplePoisDistMatrix <- as.matrix( poisd$dd )
rownames(samplePoisDistMatrix) <- dds_infected$Sample_Name
colnames(samplePoisDistMatrix) <- dds_infected$Sample_Name
```

```{r}
colors <- colorRampPalette( rev(brewer.pal(9, "RdYlBu")) )(255)
poisson_dist_plot <- pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         main = "Sample-to-Sample Poisson Distance of BL6 - Infected Vs Control",
         col = colors)
         
poisson_dist_plot
```

## PCA Plot

### Calculating all PCA Values

```{r}
 # calculate the variance for each gene
  rv <- rowVars(assay(vsd))
  ntop <- 500
  # select the ntop genes by variance
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
```

```{r}
# perform a PCA on the data in assay(x) for the selected genes
pca_infected <- prcomp(t(assay(vsd)[select,]))
summary(pca_infected)
# the contribution to the total variance for each component
percentVar_infected <- (pca_infected$sdev^2 / sum( pca_infected$sdev^2 )) * 100
percentVar_infected
```
```{r}
color_values <- c("darkred", "darkred","darkred","darkred", "black","black","darkred","darkred", "darkred",
                  "darkred", "darkred","darkred","darkred", "darkred", "darkred" ,"darkblue")
```

#### PCA Plot with VST Data

```{r}
pcaData_infected <- plotPCA(vsd, intgroup = c("condition","Sample_Name"), returnData = TRUE)
head(pcaData_infected)
percentVar_infected <- round(100 * attr(pcaData_infected, "percentVar"), digits = 3 )
percentVar_infected
```

```{r fig.height=8, fig.width=8}
PCAplot_vst <- ggplot(pcaData_infected, 
                      aes(x = PC1, 
                          y = PC2,
                          color = Sample_Name, 
                          label = Sample_Name)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar_infected[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar_infected[2], "% variance")) +
  coord_fixed() +
  scale_y_continuous(breaks = seq(-10, 10, 5)) +
  scale_x_continuous(breaks = seq(-20, 20, 5)) +
  ggtitle("PCA Plot - Infected Vs Control") +
  theme_classic() +
  geom_hline(yintercept = 0, color = "gray", size = 1) +
  geom_vline(xintercept = 0, color = "gray", size = 1) +
  theme(text = element_text(size = 15), axis.text = element_text(size = 15)) +
  theme(legend.position = "bottom") + 
  theme(aspect.ratio = 40/40) +
  geom_text(size = 4, hjust = 0, vjust = 0) +
  scale_colour_manual(values = color_values)
PCAplot_vst
ggsave(filename = '~/R/Alina_RNAseq/InfectedVsControl/PCAplot_InfectedVsControl.png',
       plot = PCAplot_vst,
       dpi = 300,
       width = 10,
       height = 10,
       units = "in")
```
## PCA Plot for different groupings of metadata

### PCA for Epithelial Response

```{r}
PCAdata_infected_ER <- plotPCA(vsd, intgroup = c("Sample_Name", "Epithelial_response"), returnData = TRUE)
head(PCAdata_infected_ER)
percentVar.PCAdata_infected_ER <- round(100 * attr(PCAdata_infected_ER, "percentVar"), digits = 4)
```

```{r}
PCAplot_ER <- ggplot(PCAdata_infected_ER,
                     aes(x = PC1, y = PC2, color = Epithelial_response, label= Sample_Name)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar.PCAdata_infected_ER[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar.PCAdata_infected_ER[2], "% variance")) +
  coord_fixed() +
  scale_y_continuous(breaks = seq(-10, 10, 5)) +
  scale_x_continuous(breaks = seq(-20, 20, 5)) +
  ggtitle("PCA Plot - Infected Vs Control - Epithelial Response") +
  theme_classic() +
  geom_hline(yintercept = 0, color = "gray", size = 1) +
  geom_vline(xintercept = 0, color = "gray", size = 1) +
  theme(text = element_text(size = 15), axis.text = element_text(size = 15)) +
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 40/40) +
  geom_text(size = 4, hjust = 0, vjust = 0) #+ scale_colour_manual(values = color_values)
PCAplot_ER
ggsave(filename = '~/R/Alina_RNAseq/InfectedVsControl/PCAplot_InfectedVsControl_EpithelialResponse.png',
       plot = PCAplot_ER,
       dpi = 300,
       width = 10,
       height = 10,
       units = "in")
```

### PCA for Microcolonies

```{r}
PCAdata_infected_MC <- plotPCA(vsd, intgroup = c("Sample_Name", "microcolonies"), returnData = TRUE)
head(PCAdata_infected_MC)
percentVar.PCAdata_infected_MC <- round(100 * attr(PCAdata_infected_MC, "percentVar"), digits = 4)
```

```{r}
PCAplot_MC <- ggplot(PCAdata_infected_MC,
                     aes(x = PC1, y = PC2,
                         color = microcolonies,
                         label= Sample_Name)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar.PCAdata_infected_MC[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar.PCAdata_infected_MC[2], "% variance")) +
  coord_fixed() +
  scale_y_continuous(breaks = seq(-10, 10, 5)) +
  scale_x_continuous(breaks = seq(-20, 20, 5)) +
  ggtitle("PCA Plot - Infected Vs Control - Microcolonies") +
  theme_classic() +
  geom_hline(yintercept = 0, color = "gray", size = 1) +
  geom_vline(xintercept = 0, color = "gray", size = 1) +
  theme(text = element_text(size = 15), axis.text = element_text(size = 15)) +
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 40/40) +
  geom_text(size = 4, hjust = 0, vjust = 0)
PCAplot_MC
ggsave(filename = '~/R/Alina_RNAseq/InfectedVsControl/PCAplot_InfectedVsControl_Microcolonies.png',
       plot = PCAplot_MC,
       dpi = 300,
       width = 10,
       height = 10,
       units = "in")
```

### PCA for Clinical Outcome

```{r}
PCAdata_infected_CO <- plotPCA(vsd, intgroup = c("Sample_Name", "clinical_outcome"), returnData = TRUE)
head(PCAdata_infected_CO)
percentVar.PCAdata_infected_CO <- round(100 * attr(PCAdata_infected_CO, "percentVar"), digits = 4)
```

```{r}
PCAplot_CO <- ggplot(PCAdata_infected_CO,
                     aes(x = PC1, y = PC2,
                         color = clinical_outcome,
                         label= Sample_Name)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar.PCAdata_infected_CO[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar.PCAdata_infected_CO[2], "% variance")) +
  coord_fixed() +
  scale_y_continuous(breaks = seq(-10, 10, 5)) +
  scale_x_continuous(breaks = seq(-20, 20, 5)) +
  ggtitle("PCA Plot - Infected Vs Control - Clinical Outcome") +
  theme_classic() +
  geom_hline(yintercept = 0, color = "gray", size = 1) +
  geom_vline(xintercept = 0, color = "gray", size = 1) +
  theme(text = element_text(size = 15), axis.text = element_text(size = 15)) +
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 40/40) +
  geom_text(size = 4, hjust = 0, vjust = 0)
PCAplot_MC
ggsave(filename = '~/R/Alina_RNAseq/InfectedVsControl/PCAplot_InfectedVsControl_ClinicalOutcome.png',
       plot = PCAplot_CO,
       dpi = 300,
       width = 10,
       height = 10,
       units = "in")
```

### PCA for ER_Microcolonies

```{r}
PCAdata_infected_ERMC <- plotPCA(vsd, intgroup = c("Sample_Name", "ER_microcolonies"), returnData = TRUE)
head(PCAdata_infected_ERMC)
percentVar.PCAdata_infected_ERMC <- round(100 * attr(PCAdata_infected_ERMC, "percentVar"), digits = 4)
```

```{r}
PCAplot_ERMC <- ggplot(PCAdata_infected_ERMC,
                       aes(x = PC1, y = PC2,
                           color = ER_microcolonies,
                           label= Sample_Name)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar.PCAdata_infected_ERMC[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar.PCAdata_infected_ERMC[2], "% variance")) +
  coord_fixed() +
  scale_y_continuous(breaks = seq(-10, 10, 5)) +
  scale_x_continuous(breaks = seq(-20, 20, 5)) +
  ggtitle("PCA Plot - Infected Vs Control - EpithelialResponse_Microcolonies") +
  theme_classic() +
  geom_hline(yintercept = 0, color = "gray", size = 1) +
  geom_vline(xintercept = 0, color = "gray", size = 1) +
  theme(text = element_text(size = 15), axis.text = element_text(size = 15)) +
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 40/40) +
  geom_text(size = 4, hjust = 0, vjust = 0)
PCAplot_ERMC
ggsave(filename = '~/R/Alina_RNAseq/InfectedVsControl/PCAplot_InfectedVsControl_ER_Microcolonies.png',
       plot = PCAplot_ERMC,
       dpi = 300,
       width = 10,
       height = 10,
       units = "in")
```

### PCA for Phylogenomic Lineage

```{r}
PCAdata_infected_PL <- plotPCA(vsd, intgroup = c("Sample_Name", "phylogenomic_lineage"), returnData = TRUE)
head(PCAdata_infected_PL)
percentVar.PCAdata_infected_PL <- round(100 * attr(PCAdata_infected_PL, "percentVar"), digits = 4)
```

```{r}
PCAplot_PL <- ggplot(PCAdata_infected_PL,
                       aes(x = PC1, y = PC2,
                           color = phylogenomic_lineage,
                           label= Sample_Name)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar.PCAdata_infected_PL[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar.PCAdata_infected_PL[2], "% variance")) +
  coord_fixed() +
  scale_y_continuous(breaks = seq(-10, 10, 5)) +
  scale_x_continuous(breaks = seq(-20, 20, 5)) +
  ggtitle("PCA Plot - Infected Vs Control - Phylogenomic Lineage") +
  theme_classic() +
  geom_hline(yintercept = 0, color = "gray", size = 1) +
  geom_vline(xintercept = 0, color = "gray", size = 1) +
  theme(text = element_text(size = 15), axis.text = element_text(size = 15)) +
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 40/40) +
  geom_text(size = 4, hjust = 0, vjust = 0)
PCAplot_PL
ggsave(filename = '~/R/Alina_RNAseq/InfectedVsControl/PCAplot_InfectedVsControl_PhylogenomicLineage.png',
       plot = PCAplot_PL,
       dpi = 300,
       width = 10,
       height = 10,
       units = "in")
```
### PCA for Phylogroup

```{r}
PCAdata_infected_PG <- plotPCA(vsd, intgroup = c("Sample_Name", "phylogroup"), returnData = TRUE)
head(PCAdata_infected_PG)
percentVar.PCAdata_infected_PG <- round(100 * attr(PCAdata_infected_PG, "percentVar"), digits = 4)
```

```{r}
PCAplot_PG <- ggplot(PCAdata_infected_PG,
                     aes(x = PC1, y = PC2,
                         color = phylogroup,
                         label= Sample_Name)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar.PCAdata_infected_PG[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar.PCAdata_infected_PG[2], "% variance")) +
  coord_fixed() +
  scale_y_continuous(breaks = seq(-10, 10, 5)) +
  scale_x_continuous(breaks = seq(-20, 20, 5)) +
  ggtitle("PCA Plot - Infected Vs Control - Phylogroup") +
  theme_classic() +
  geom_hline(yintercept = 0, color = "gray", size = 1) +
  geom_vline(xintercept = 0, color = "gray", size = 1) +
  theme(text = element_text(size = 15), axis.text = element_text(size = 15)) +
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 40/40) +
  geom_text(size = 4, hjust = 0, vjust = 0)
PCAplot_PG
ggsave(filename = '~/R/Alina_RNAseq/InfectedVsControl/PCAplot_InfectedVsControl_Phylogroup.png',
       plot = PCAplot_PG,
       dpi = 300,
       width = 10,
       height = 10,
       units = "in")
```

### PCA for Intimin Group

```{r}
PCAdata_infected_IT <- plotPCA(vsd, intgroup = c("Sample_Name", "Intimin_Type"), returnData = TRUE)
head(PCAdata_infected_IT)
percentVar.PCAdata_infected_IT <- round(100 * attr(PCAdata_infected_IT, "percentVar"), digits = 4)
```

```{r}
PCAplot_IT <- ggplot(PCAdata_infected_IT,
                     aes(x = PC1, y = PC2,
                         color = Intimin_Type,
                         label= Sample_Name)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar.PCAdata_infected_IT[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar.PCAdata_infected_IT[2], "% variance")) +
  coord_fixed() +
  scale_y_continuous(breaks = seq(-10, 10, 5)) +
  scale_x_continuous(breaks = seq(-20, 20, 5)) +
  ggtitle("PCA Plot - Infected Vs Control - Intimin Type") +
  theme_classic() +
  geom_hline(yintercept = 0, color = "gray", size = 1) +
  geom_vline(xintercept = 0, color = "gray", size = 1) +
  theme(text = element_text(size = 15), axis.text = element_text(size = 15)) +
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 40/40) +
  geom_text(size = 4, hjust = 0, vjust = 0)
PCAplot_IT
ggsave(filename = '~/R/Alina_RNAseq/InfectedVsControl/PCAplot_InfectedVsControl_IntiminType.png',
       plot = PCAplot_IT,
       dpi = 300,
       width = 10,
       height = 10,
       units = "in")
```

## Hierarchical Clustering

### applying rlog Transformation

```{r}
rld <- rlog(dds_infected, blind = FALSE)
head(assay(rld), 3)
```
```{r}
### Extract the rlog matrix from the object
rld_mat <- assay(rld) #assay() is function from the "SummarizedExperiment" package that was loaded when you loaded DESeq2
```

```{r}
### Compute pairwise correlation values
rld_cor <- cor(rld_mat)    ## cor() is a base R function
head(rld_cor)   ## check the output of cor(), make note of the rownames and colnames
```

```{r}
### Plot heatmap
heat.colors <- brewer.pal(6, "RdYlBu")
Hclust_plot <- pheatmap(rld_cor, 
                        color = heat.colors,
                        main = 'Heirarchical Clustering of Samples - InfectedVsControl - Correlation Matrix'
                        #filename = '~/R/Alina_RNAseq/InfectedVsControl/Hclust_plot.tiff'
                       )
Hclust_plot
```

## DGE Results

### Running the differential expression pipeline

```{r}
dds1_infected <- DESeq(dds_infected)
#str(dds1)
```

### Building the results table

```{r}
res_infected <- results(dds1_infected)
head(res_infected, 30)
```

```{r}
summary(res_infected)
```

A Problem: there a number of rows with padj that has NA.

Reaons:

1.  If within a row, all samples have zero counts, the baseMean column
    will be zero, and the log2 fold change estimates, p value and
    adjusted p value will all be set to NA. ------\>\>\>\>\>\> checked
    and not true!

2.  If a row contains a sample with an extreme count outlier then the p
    value and adjusted p value will be set to NA. These outlier counts
    are detected by Cook's distance. Customization of this outlier
    filtering and description of functionality for replacement of
    outlier counts and refitting is described below. ------\>\>\>\>\>\>
    checked and not true!

3.  If a row is filtered by automatic independent filtering, for having
    a low mean normalized count, then only the adjusted p value will be
    set to NA. Description and customization of independent filtering is
    described below. ------\>\>\>\>\>\> could be true!

=====\>\>\>\>\>\> Most likely caused by automatic independent filtering
due to the presence of low mean normalized counts.

Solution: Obtain unfiltered DESeq2 results! (counts without automatic
independent filtering and outlier removal). Then, only the genes with
ALL counts set to zero will have NA for pvalues.

```{r}
dds2_infected <- DESeq(dds_infected, minReplicatesForReplace=Inf)
res2_infected <- results(dds2_infected, cooksCutoff=FALSE, independentFiltering=FALSE)
```

```{r}
head(res2_infected, 30)
```

### Results with thresholds

Both thresholds Applied!

```{r}
res2df_infected <- as.data.frame(res2_infected) # convert the results table to a df
head(res2df_infected)
```

## MA Plot

```{r}
resultsNames(dds2_infected)
```

```{r}
plotMA_res2_infected <- plotMA(res2_infected, ylim = c(-2, 2))
```

```{r}
res3_infected <- lfcShrink(dds2_infected, coef="condition_Infected_vs_control", type="apeglm")
plotMA_res3 <-plotMA(res3_infected, ylim = c(-2, 2))
```

### Histogram of p-values

```{r}
hist(res2_infected$pvalue, breaks = 50, col = "grey50", border = "blue")
```

Further Filtering: baseMean \> 1

```{r}
hist(res2_infected$pvalue[res2_infected$baseMean > 1], breaks = 50, col = "grey50", border = "blue")
```

## Annotating and Exporting Results

-   adding gene annotation to results table
-   adding ENTREZ Id to results table

```{r}
res2df_infected$symbol <- mapIds(org.Mm.eg.db,
                     keys = rownames(dds_infected),
                    column = "SYMBOL",
                    keytype = "ENSEMBL",
                     multiVals = "first")
res2df_infected$entrez <- mapIds(org.Mm.eg.db,
                     keys = rownames(dds_infected),
                    column = "ENTREZID",
                    keytype = "ENSEMBL",
                     multiVals = "first")
```

```{r}
head(res2df_infected)
```

```{r}
str(res2df_infected)
nrow(res2df_infected)
```

Omit NA values from symbol and respective rows!

```{r}
res3df_infected <- res2df_infected %>% filter(!is.na(symbol) & !is.na(entrez))
```
```{r}
nrow(res3df_infected)
```

## Saving the Results

```{r}
resOrdered_infected <- res3df_infected[order(res3df_infected$pvalue),]
head(resOrdered_infected)
```

```{r}
write.csv(as.data.frame(resOrdered_infected),
          file = "~/R/Alina_RNAseq/InfectedVsControl/results_DGE_InfectedVsControl.csv")
```

## Heatmap of count matrix

To explore a count matrix, it is often instructive to look at it as a heatmap.

```{r}
select <- order(rowMeans(counts(dds2_infected,normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds2_infected)[,c("condition","Sample_Name")])
```

```{r}
pheatmap(assay(vsd)[select,], 
         cluster_cols = TRUE, 
         annotation_col = df,
         color = heat.colors,
         show_rownames = FALSE)
```
## Effect of Transformations on Variance

-   These set of plots depict the standard deviation of transformed data (across samples), against mean.

### Based on Shifted Log Transformation

```{r}
ntd <- normTransform(dds2_infected)
meanSdPlot(assay(ntd))
```

### Based on Rlog Transformation

```{r}
meanSdPlot(assay(rld))
```

### Based on Variance Stabilizing Transformation

```{r}
meanSdPlot(assay(vsd))
```

## Dispersion Plots

-   Its a useful diagnostic to plot the dispersion estimates.

```{r}
plotDispEsts(dds2_infected)
```

# Volcano Plots

## Volcano Plots based on Enhanced Volcano

```{r}
p1 <- EnhancedVolcano(res3df_infected,
    lab = res3df_infected$symbol,
    x = 'log2FoldChange',
    y = 'pvalue',
    xlab = bquote(~Log[2]~'FoldChange'),
    pCutoff = 0.05,
    FCcutoff = 1.0,
    title = 'Volcano Plot for DE genes: Log2FoldChange Vs -Log10pValue',
    pointSize = 2.0,
    labSize = 5.0,
    boxedLabels = FALSE,
    gridlines.major = FALSE,
    gridlines.minor = FALSE,
    colAlpha = 0.5,
    xlim = c(-6, 9),
    ylim = c(-2, 12),
    legendPosition = 'bottom',
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.75
    )
p1 +
  scale_y_continuous(limits = c(-1, 8) ,breaks = c(-1,0,1,2,3,4,5,6,7,8)) 
  scale_x_continuous(limits = c(-6, 9) , breaks = seq(-6, 9, 1))
```

## volcano Plot of Log2FC vs Padj

```{r}
p2 <- EnhancedVolcano(res3df_infected,
    lab = res3df_infected$symbol,
    x = 'log2FoldChange',
    y = 'padj',
    xlab = bquote(~Log[2]~'FoldChange'),
    ylab = bquote(~Log[10]~'padj'),
    pCutoff = 0.05,
    FCcutoff = 1.0,
    title = 'Volcano Plot for DE genes: Log2FoldChange Vs -Log10padj',
    pointSize = 2.0,
    labSize = 5.0,
    boxedLabels = FALSE,
    gridlines.major = FALSE,
    gridlines.minor = FALSE,
    colAlpha = 0.5,
    xlim = c(-6, 9),
    ylim = c(-2, 12),
    legendPosition = 'bottom',
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.75
    )
p2 +
  scale_y_continuous(limits = c(-1, 8) ,breaks = c(-1,0,1,2,3,4,5,6,7,8)) 
  scale_x_continuous(limits = c(-6, 9) , breaks = seq(-6, 9, 1))
```

## Significant Differentially Expressed Genes

```{r}
res3df_infected$diffexpressed <- "NS"
# if log2Foldchange > 1.0 and pvalue < 0.05, set as "UP"
res3df_infected$diffexpressed[res3df_infected$log2FoldChange > 1.0 & res3df_infected$pvalue < 0.05] <- "UP"
# if log2Foldchange < 1.0 and pvalue < 0.05, set as "UP"
res3df_infected$diffexpressed[res3df_infected$log2FoldChange < -1.0 & res3df_infected$pvalue < 0.05] <- "DOWN"
# Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
res3df_infected$delabel <- NA
res3df_infected$delabel[res3df_infected$diffexpressed != "NS"] <- res3df_infected$symbol[res3df_infected$diffexpressed != "NS"]
```

```{r}
#ggplot(res3.df, aes(x=log2FoldChange , y = log10(baseMean), 
#                    col=diffexpressed, label=delabel)
#       ) + 
#  geom_point() +
#  scale_color_manual(values = mycolors)+
#  geom_text_repel()
```

Arrive at relevant genes by imposing thresholds.

```{r}
sigsdf_infected <- res3df_infected[(abs(res3df_infected$log2FoldChange)>1) & (res3df_infected$pvalue< 0.05),]
```
```{r}
nrow(sigsdf_infected)
```
## Log2FC Table

#by_gene_forlog2FC \<- UpGeneFrame_Core50 %\>% group_by(symbol) %\>%
summarise(log2FoldChange, Strain_name) %\>% ungroup()
head(by_gene_forlog2FC)

```{r}
summary(res3df_infected)
```

### Gender Genes Removed from Table

```{r}
sigsdf_infected <- filter(sigsdf_infected, 
                          symbol != "Xist",
                          symbol != "Jpx", 
                          symbol != "Ftx", 
                          symbol != "Tsx",
                          symbol != "Cnbp2" )
```

```{r}
nrow(sigsdf_infected)
#head(sigsdf_infected)
```

Therefore, gender genes werent so much in play in terms of significance!

```{r}

write.csv(sigsdf_infected ,"~/R/Alina_RNAseq/InfectedVsControl/SignificantGenes_InfectedVsControl.csv")
```

### Number of Genes from different strains that are contributing to UP/DOWN regulation.

```{r}
sigsdf_infected_UP <- sigsdf_infected[(sigsdf_infected$log2FoldChange) > 1, ] #UP Regulation Table
sigsdf_infected_DOWN <- sigsdf_infected[(sigsdf_infected$log2FoldChange) < -1, ] #DOWN Regulation Table
```

```{r}
nrow(sigsdf_infected_UP)
nrow(sigsdf_infected_DOWN)
```

### Determine Top20 UP Genes and Top20 DOWN Genes

```{r}
UpGene_Infected <- sigsdf_infected_UP[order(-sigsdf_infected_UP$log2FoldChange), ]$symbol
DownGene_Infected <- sigsdf_infected_DOWN[order(-sigsdf_infected_DOWN$log2FoldChange), ]$symbol
DE_Genes_table_Infected <- as.data.frame(cbind(UpGene_Infected,DownGene_Infected)) # sorted based on highest Log2FC value
```

```{r}
write.csv(DE_Genes_table_Infected ,"~/R/Alina_RNAseq/DE_GenesTable_InfectedVsControl.csv")
head(DE_Genes_table_Infected, 20)
```

## Z-score based Gene Heatmaps

```{r}
head(sigsdf_infected)
```

### with Whole table (all genes together!)

```{r}
mat <- counts(dds2_infected, normalized = TRUE)[rownames(sigsdf_infected),]
mat.zs <- t(apply(mat, 1, scale)) # Calculating the zscore for each row
colnames(mat.zs) <- coldata$Sample_Name # need to provide correct sample names for each of the columns
head(mat.zs)
```

```{r}
Heatmap_ALL_DEGene <- pheatmap(mat.zs,
                               cluster_cols = TRUE,
                               cluster_rows = FALSE,
                               show_rownames = FALSE)
Heatmap_ALL_DEGene
```

```{r}
sigs2.df <- res2.df[(abs(res2.df$log2FoldChange)>1) & (res2.df$pvalue< 0.05),]
mat2 <- counts(dds2, normalized = TRUE)[rownames(sigs2.df),]
mat2.zs <- t(apply(mat2, 1, scale)) # Calculating the zscore for each row
colnames(mat2.zs) <- coldata$Samples # need to provide correct sample names for each of the columns
head(mat2.zs)
#pheatmap(mat2.zs, cluster_cols = TRUE, cluster_rows = FALSE, show_rownames = FALSE)
```

```{r fig.height= 15, fig.width= 15}
newHP <- Heatmap(mat2.zs, 
        cluster_columns = TRUE, 
        cluster_rows = TRUE, 
        column_labels = colnames(mat2.zs),
        name = 'Z-Score Heatmap of DE Genes',
        show_row_names = FALSE, 
        use_raster = TRUE, 
        raster_quality = 5,
        #width = unit(8, "in"), 
        #height = unit(8, "in")
        row_labels = sigs2.df[rownames(mat2.zs),]$symbol
        )
newHP
```

```{r fig.height= 75, fig.width= 10}
newHP1 <- Heatmap(mat2.zs, 
        cluster_columns = TRUE, 
        cluster_rows = TRUE, 
        column_labels = colnames(mat2.zs),
        row_labels = sigs2.df[rownames(mat2.zs),]$symbol,
        name = 'Z-Score Heatmap of DE Genes',
        show_row_names = TRUE, 
        use_raster = TRUE, 
        raster_quality = 5,
        #width = unit(10, "in"), 
        #height = unit(45, "in")
        )
newHP1
```

------------------------------------------------------------------------

Need to filter these results to accommodate better the heat maps and
also volcano plots

### with tighter constraints (all genes together!)

```{r}
sigs1.df <- res2.df[(res2.df$baseMean > 75) & (abs(res2.df$log2FoldChange)>2) & (res2.df$pvalue< 0.05),]
mat1 <- counts(dds2, normalized = TRUE)[rownames(sigs1.df),]
mat1.zs <- t(apply(mat1, 1, scale)) # Calculating the zscore for each row
colnames(mat1.zs) <- coldata$Samples # need to provide correct sample names for each of the columns
head(mat1.zs)
```

```{r}
Heatmap(mat1.zs,
        cluster_columns = TRUE,
        cluster_rows = TRUE,
        column_labels = colnames(mat1.zs),
        name = 'Z-Score Heatmap of DE Genes',
        row_labels = sigs1.df[rownames(mat1.zs),]$symbol)
```

# GO Terms with clusterProfiler

## GO Terms for UP Regulated Genes

### GO over-representation analysis for UP Regulated Genes

```{r}
UPgene_ENS_ID <- rownames(sigs.UP.df)
```

```{r}
GO_UPReg_results <- enrichGO(gene = UPgene_ENS_ID,
                       OrgDb = "org.Mm.eg.db",
                       keyType = "ENSEMBL",
                       ont = "BP",
                       pAdjustMethod = "BH",
                       pvalueCutoff  = 0.01,
                       qvalueCutoff  = 0.05,
                       readable      = TRUE
                       )
GO.UpReg.df <- as.data.frame(GO_UPReg_results)
GO.UpReg.df
```

```{r}
GO_UPReg_results
```

```{r}
write.csv(GO.UpReg.df ,"~/R/Alina_RNAseq/GO.UpReg.df_results.csv")
```

```{r fig.height=10, fig.width=10}
GO_fit_UPReg <- plot(barplot(GO_UPReg_results, showCategory = 22))
```

```{r}
ggsave(filename = '~/R/Alina_RNAseq/GO_fit_UPReg_barplot.tiff', 
       plot = GO_fit_UPReg, 
       dpi = 300,
       width = 12,
       height = 10,
       units = "in")
```

```{r fig.height=10, fig.width=10}
GO_fit_UPReg_dot <- plot(dotplot(GO_UPReg_results, showCategory = 23))
```

```{r}
ggsave(filename = '~/R/Alina_RNAseq/GO_fit_UPReg_dotplot.tiff', 
       plot = GO_fit_UPReg_dot, 
       dpi = 300,
       width = 12,
       height = 10,
       units = "in")
```

```{r fig.height=10, fig.width=10}
GO_fit_UPReg_cnet <- plot(cnetplot(GO_UPReg_results, showCategory = 13))
```

```{r}
ggsave(filename = '~/R/Alina_RNAseq/GOfi_UPReg_cnetplot.tiff', 
       plot = GO_fit_UPReg_cnet, 
       dpi = 300,
       width = 12,
       height = 10,
       units = "in")
```

```{r fig.height=20, fig.width=10}
d <- godata('org.Mm.eg.db', ont="BP")
ego2 <- pairwise_termsim(GO_UPReg_results, method="Wang", semData = d)
emapplot(ego2, showCategory = 30)
```

## GO Terms for Down Regulated Genes

### GO over-representation analysis for DOWN Regulated Genes

```{r}
DOWNgene_ENS_ID <- rownames(sigs.DOWN.df)
```

```{r}
#DOWNgene_ENS_ID
```

```{r}
GO_DOWNReg_results <- enrichGO(gene = DOWNgene_ENS_ID,
                       OrgDb = "org.Mm.eg.db",
                       keyType = "ENSEMBL",
                       ont = "BP",
                       pAdjustMethod = "BH",
                       pvalueCutoff  = 0.01,
                       qvalueCutoff  = 0.05,
                       readable      = TRUE
                       )
GO_DOWNReg_results
```

```{r}
sessionInfo( )
```
```{r}