---
title: "Differential Gene Expression Analysis Notebook - Alina's Data Analysis Project"
output: 
  html_document: 
    toc: yes
    number_sections: yes
---

                                             
The DGE analysis will be performed using the raw integer read counts for control and infected conditions.
The goal here is to identify the differentially expressed genes under infected condition.

```{r}
library("DESeq2")
library("PoiClaClu")
library("pheatmap")
library("RColorBrewer")
```

```{r}
cm1 <- as.matrix(read.csv("C:/Users/kesha/Documents/R/Rtuts/Data/WG__Discussion_Alina's_EPEC_project/fc/counts.csv", row.names = "X"))
dim(cm1)
```

DESeq2 needs sample information (metadata) for performing DGE analysis. Letâ€™s create the sample information (you can also import sample information if you have it in a file).

```{r}
sample_ID <- list(read.csv("C:/Users/kesha/Documents/R/Rtuts/Data/WG__Discussion_Alina's_EPEC_project/fc/samples.csv", row.names = "X"))
sample_ID
```
```{r}
condition = c("Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","control","control")
```


```{r}
coldata <- data.frame(sample_ID, condition)
colnames(coldata) <- c('sample','condition') # change name of one of the columns
coldata$condition <- as.factor(coldata$condition)
coldata <- data.frame(coldata, row.names = 1) # convert column to row.names
coldata 
```

It is essential to have the name of the columns in the count matrix in the same order as that in name of the samples (rownames in coldata)

```{r}
all(rownames(coldata) %in% colnames(cm1))
```
Now, construct DESeqDataSet for DGE analysis:


```{r}
dds <- DESeqDataSetFromMatrix(countData = cm1, 
                              colData = coldata, 
                              design = ~ condition)
```

```{r}
dds
```


Pre-filter the genes which have low counts. Here, I will remove the genes which have < 10 reads (this can vary based on research goal) in total across all the samples. Pre-filtering helps to remove genes that have very few mapped reads, reduces memory, and increases the speed of the DESeq2 analysis.

```{r}
dds <- dds[rowSums(counts(dds)) >= 5,]
```

Now, select the reference level for condition comparisons. The reference level can set using ref parameter. The comparisons of other conditions will be compared against this reference i.e, the log2 fold changes will be calculated based on ref value (infected/control) . If this parameter is not set, comparisons will be based on alphabetical order of the levels.

```{r}
dds$condition <- relevel(dds$condition, ref = "control")
```

Perform Differntial Gene Expression Analysis:
```{r}
dds1 <- DESeq(dds)
```
# get gene expression table

```{r}
resultsNames(dds1)
```


```{r}
res <- results(dds1, independentFiltering=FALSE, name = "condition_Infected_vs_control")
res
```

Order gene expression table by adjusted p value (Benjamini-Hochberg FDR method)
```{r}
res[order(res$padj),]  
```


Note on NA:  some genes with p value set to NA. This is due to all samples have zero counts for a gene or there is extreme outlier count for a gene or that gene is subjected to independent filtering by DESeq2.



```{r}
write.csv(as.data.frame(res[order(res$padj),] ), 
          file="C:/Users/kesha/Documents/R/Rtuts/Data/WG__Discussion_Alina's_EPEC_project/results/condition_infected_vs_control_dge.csv")
```

```{r}
summary(results(dds1, alpha=0.05))
```


```{r}
summary(results(dds1, alpha=0.05, lfcThreshold =1))
```

```{r}
sizeFactors(dds1)
```

# Explaratory Data analysis

```{r}
plotMA(res)
```



```{r}
plotDispEsts(dds1)
```
# Heatmap of sample-to-sample distances using the variance stabilizing transformed values.


```{r}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- names(vsd$sizeFactor)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists
         )
```


```{r}

poisd <- PoissonDistance(t(counts(dds1)))
samplePoisDistMatrix <- as.matrix( poisd$dd )
rownames(samplePoisDistMatrix) <- names(dds1$sizeFactor)
colnames(samplePoisDistMatrix) <- NULL
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd
         )
```

