---
title: "Differential Gene Expression Analysis Notebook - Alina's Data Analysis Project"
output: 
  html_document: 
    toc: yes
    number_sections: yes
---

                                             
The DGE analysis will be performed using the raw integer read counts for control and infected conditions.
The goal here is to identify the differentially expressed genes under infected condition.

```{r}
library("DESeq2")
library("PoiClaClu")
library("pheatmap")
library("RColorBrewer")
library(tidyverse)
```

```{r}
cm1 <- as.matrix(read.csv("C:/Users/kesha/Documents/R/Rtuts/Data/WG__Discussion_Alina's_EPEC_project/fc/counts.csv", row.names = "X"))
dim(cm1)
```


```{r}
head(cm1, 10)
```


DESeq2 needs sample information (metadata) for performing DGE analysis. Letâ€™s create the sample information (you can also import sample information if you have it in a file).

```{r}
sample_ID <- list(read.csv("C:/Users/kesha/Documents/R/Rtuts/Data/WG__Discussion_Alina's_EPEC_project/fc/samples.csv", row.names = "X"))
sample_ID
```
```{r}
condition = c("Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","Infected","control","control")
```


```{r}
coldata <- data.frame(sample_ID, condition)
colnames(coldata) <- c('sample','condition') # change name of one of the columns
coldata
```


```{r}
coldata$condition <- as.factor(coldata$condition)
coldata <- data.frame(coldata, row.names = 1) # convert column to row.names
coldata 
```

It is essential to have the name of the columns in the count matrix in the same order as that in name of the samples (rownames in coldata)

```{r}
all(rownames(coldata) %in% colnames(cm1))
```
Now, construct DESeqDataSet for DGE analysis:


```{r}
dds <- DESeqDataSetFromMatrix(countData = cm1, 
                              colData = coldata, 
                              design = ~ condition)
```

```{r}
dds
```

# Normalization


## Count Normalization via DESq2

```{r}
dds <- estimateSizeFactors(dds)
```

Can check the normalization factors using:
```{r}
sizeFactors(dds)
```

```{r}
normalized_counts <- counts(dds, normalized = TRUE)
```

Saving the normalized counts table.
```{r}
write.table(normalized_counts, 
          file="C:/Users/kesha/Documents/R/Rtuts/Data/WG__Discussion_Alina's_EPEC_project/results/normalized_counts.txt",
          sep = "\t",
          quote = FALSE,
          col.names = NA)
```

ASK Q1!


# Quality Control


Assess overall similarity between samples and understand the following:

1. Which samples are similar to each other and which are different?
2. Does this fit to the expectation from experiment's design?
3. What are the major sources of variation in dataset?

## Count Data Tansformations - Extracting Transformed Values


### Transform normalised counts using rlog transformation for Visualization (only!)

```{r}
rld <- rlog(dds, blind = TRUE) 
# blind flag turned on to keep the transformation unbiased to sample condition information
vsd <- vst(dds, blind = TRUE)

```


```{r}
head(assay(vsd), 3)
```

### Effects of transformations on the variance


```{r}
# this gives log2(n + 1)
ntd <- normTransform(dds)
library("vsn")
meanSdPlot(assay(ntd))
```

```{r}
meanSdPlot(assay(vsd))
```
```{r}
meanSdPlot(assay(rld))
```
## Data Quality assessment via sample clustering and visualization


### Heatmap of Count Matrix

```{r}

select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds))
```


```{r}

pheatmap(assay(ntd)[select,], cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=df)
```
```{r}
pheatmap(assay(vsd)[select,], cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=df)
```



### Heatmap of sample-to-sample distances using the variance stabilizing transformed values.


```{r}
sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- names(vsd$sizeFactor)
colnames(sampleDistMatrix) <- NULL
```


```{r}

colors <- colorRampPalette( rev(brewer.pal(9, "Spectral")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col= colors
         )
```
### Plot PCA of Samples

```{r}
plotPCA(vsd, intgroup="condition")
```


### Hierarchical Clustering

```{r}
### Extract the rlog matrix from the object
rld_mat <- assay(rld)    ## assay() is function from the "SummarizedExperiment" package that was loaded when you loaded DESeq2
```


```{r}
### Compute pairwise correlation values
rld_cor <- cor(rld_mat)    ## cor() is a base R function

head(rld_cor)   ## check the output of cor(), make note of the rownames and colnames
```

```{r}
### Plot heatmap
pheatmap(rld_cor, border_color=NA, 
         fontsize = 5, 
         fontsize_row = 10, 
         height=50)
```
```{r}
heat.colors <- brewer.pal(6, "Blues")
pheatmap(rld_cor, color = heat.colors, border_color=NA, fontsize = 7, 
			fontsize_row = 10, height=50)
```


# DGE Analysis

Pre-filter the genes which have low counts. Here, I will remove the genes which have < 10 reads (this can vary based on research goal) in total across all the samples. Pre-filtering helps to remove genes that have very few mapped reads, reduces memory, and increases the speed of the DESeq2 analysis.

```{r}
dds <- dds[rowSums(counts(dds)) >= 5,]
```

Now, select the reference level for condition comparisons. The reference level can set using ref parameter. The comparisons of other conditions will be compared against this reference i.e, the log2 fold changes will be calculated based on ref value (infected/control) . If this parameter is not set, comparisons will be based on alphabetical order of the levels.

```{r}
dds$condition <- relevel(dds$condition, ref = "control")
```

Perform Differntial Gene Expression Analysis:
```{r}
dds1 <- DESeq(dds)
```
# get gene expression table

```{r}
resultsNames(dds1)
```


```{r}
res <- results(dds1, independentFiltering=FALSE, name = "condition_Infected_vs_control")
res
```

Order gene expression table by adjusted p value (Benjamini-Hochberg FDR method)
```{r}
res[order(res$padj),]  
```


Note on NA:  some genes with p value set to NA. This is due to all samples have zero counts for a gene or there is extreme outlier count for a gene or that gene is subjected to independent filtering by DESeq2.



```{r}
write.csv(as.data.frame(res[order(res$padj),] ), 
          file="C:/Users/kesha/Documents/R/Rtuts/Data/WG__Discussion_Alina's_EPEC_project/results/condition_infected_vs_control_dge.csv")
```

```{r}
summary(results(dds1, alpha=0.05))
```


```{r}
summary(results(dds1, alpha=0.05, lfcThreshold =1))
```

```{r}
sizeFactors(dds1)
```

# Explaratory Data analysis

```{r}
plotMA(res)
```



```{r}
plotDispEsts(dds1)
```





