---
title: "From Bams to Count Matrix and then DGE - Notebook"
output:
  html_document:
    toc: true
---


```{r}

library('Rsubread')
```

# Feature counts
```{r}
# GTF File
gtf_file <- "D:/3primeRNASeqData_fastq_bam_countmatrices/Mus_musculus.GRCm39.106.chr.gtf"

```

```{r}
# Bam files
bam_path <- "D:/3primeRNASeqData_fastq_bam_countmatrices/mixture/"
bamFiles <- list.files(bam_path, pattern = ".*bam$")
bamFiles
```


```{r}
setwd(bam_path)
fc <- featureCounts(files = bamFiles,
                    annot.ext= gtf_file,
                    isGTFAnnotationFile=TRUE,
                    isPairedEnd=FALSE)
```



```{r}
head(fc)
```
```{r}
head(fc[["annotation"]], 10)

```

```{r}
head(fc[["counts"]], 10)
write.csv(fc[["counts"]],
          'D:/3primeRNASeqData_fastq_bam_countmatrices/mixture/mixturecounts.csv')
```
